src/util/invariant.h

```diff
@@ -81,24 +133,83 @@ class invariant_failedt: public std::logic_error
 
 
 #else
-// CPROVER_INVARIANT_PRINT_STACK_TRACE affects the implementation of
-// this function but not it's generation from the macro
 
-void check_invariant(
-  const char * const file,
-  const char * const function,
+void print_backtrace(
+  const std::string &file,
+  const std::string &function,
+  const int line,
+  std::ostream &out);
+
+std::string get_backtrace(
+  const std::string &file,
+  const std::string &function,
+  const int line);
+
+void report_exception_to_stderr(const invariant_failedt &);
+
+/// Takes a backtrace, gives it to the reason structure, then aborts, printing
+/// reason.what() (which therefore includes the backtrace).
+/// In future this may throw `reason` instead of aborting.
+/// \param file : C string giving the name of the file.
+/// \param function : C string giving the name of the function.
+/// \param line : The line number of the invariant
+/// \param reason : (derived from invariant_failedt) error report structure;
+///  its backtrace member will be set before it is used.
+template<class T>
+typename std::enable_if<std::is_base_of<invariant_failedt, T>::value>::type
+invariant_violated(
+  const std::string &file,
+  const std::string &function,
   const int line,
-  const bool condition,
-  const char * const reason);
+  T reason)
+{
+  reason.backtrace=get_backtrace(file, function, line);
+  // We now have a structured exception ready to use;
+  // in future this is the place to put a 'throw'.
+  report_exception_to_stderr(reason);
+  abort();
+}
+
+/// Takes a backtrace, constructs an invariant_violatedt from reason and the
+/// backtrace, aborts printing the invariant's description.
+/// In future this may throw rather than aborting.
+/// \param file : C string giving the name of the file.
+/// \param function : C string giving the name of the function.
+/// \param line : The line number of the invariant
+/// \param reason : (std::string or const char*) brief description
+///                 of the invariant violation.
+template<class T>
+typename std::enable_if<std::is_constructible<std::string, T>::value>::type
+invariant_violated(
+  const std::string &file,
+  const std::string &function,
+  const int line,
+  const T &reason)
```

Comment:
This is going to be passed as a `const std::string &what` to the `invariant_failedt` constructor so might as well be passed as that here, which will simplify things. I believe it should still be able to sit alongside your templated overload above if it isn't templated here.